src/Node.cc

```diff
@@ -526,25 +566,44 @@ bool Node::Publisher::Publish(const ProtoMsg &_msg)
   }
 
   // Handle remote subscribers.
-  if (subscribers.haveRemote)
+  std::string impl = this->dataPtr->shared->GzImplementation();
+  if (impl == "zeromq")
   {
-    // Zmq will call this lambda when the message is published.
-    // We use it to deallocate the buffer.
-    auto myDeallocator = [](void *_buffer, void *)
+    if (subscribers.haveRemote)
     {
-      delete[] reinterpret_cast<char*>(_buffer);
-    };
+      // Zmq will call this lambda when the message is published.
+      // We use it to deallocate the buffer.
+      auto myDeallocator = [](void *_buffer, void *)
+      {
+        delete[] reinterpret_cast<char*>(_buffer);
+      };
 
-    if (!this->dataPtr->shared->Publish(this->dataPtr->publisher.Topic(),
-          msgBuffer, msgSize, myDeallocator, std::string(_msg.GetTypeName())))
+      if (!this->dataPtr->shared->Publish(this->dataPtr->publisher.Topic(),
+            msgBuffer, msgSize, myDeallocator, std::string(_msg.GetTypeName())))
+      {
+        return false;
+      }
+    }
+    else
     {
-      return false;
+      delete[] msgBuffer;
     }
   }
-  else
+#ifdef HAVE_ZENOH
+  else if (impl == "zenoh")
   {
-    delete[] msgBuffer;
+    if (subscribers.haveRemote)
+    {
+      zenoh::Publisher::PutOptions options;
+      // Add message type as an attachment.
+      options.attachment = this->dataPtr->publisher.MsgTypeName();
+
+      this->dataPtr->zPub->put(msgBuffer, std::move(options));
```

Comment:
I was testing this with gz-sim and ran into error messages that said 

```
SubscriptionHandler::CreateMsg() error: ParseFromString failed
```

After investigating a bit, I pinpointed the problem to this line. What I believe is happening here is that `zenoh::Bytes` copies `msgBuffer` as though it was. a c-string, so if a null character happens to be in the message, the resulting message published would be truncated. The solution here is to use the constructor that takes a size argument which takes ownership of the buffer instead of copying it and relies on the provided size, so null characters are not a problem. After making this change and my suggested change above, I was able to run `gz sim shapes.sdf` succesfully 🎉 


```suggestion
      // Zenoh will call this lambda once Bytes objects are destroyed
      auto deleter = [](uint8_t *_buffer) { delete[] _buffer; };
      auto zMsgBuffer = reinterpret_cast<uint8_t *>(msgBuffer);
      this->dataPtr->zPub->put(zenoh::Bytes(zMsgBuffer, msgSize, deleter),
                               std::move(options));

```